### PR TITLE
[sglang][lora] Support DP attention in LoRA backends

### PR DESCRIPTION
**Pull request by erikwijmans**
## Motivation

DP Attention + LoRA is currently unsupported. It would be nice to support this.

## Modifications

- Add `LoRABatchLayout` to select DP-local or TP-global LoRA routing explicitly, including at communicator transitions and in DeepSeek-family models.
- Keep DP-local LM-head pruning local, and gather pruned LM-head routes only for the TP-global LM head.
- Reuse padded model routing for decode/verify CUDA graphs without carrying extend/logprob state into replay views. LoRA prefill remains eager with DP attention.
- Require GPU-pinned adapters with DP attention; retain the existing limit of `max_loras_per_batch - 1` pinned adapters. No change to non-DP admission or residency behavior.
- Aggregate dynamic adapter load/unload results across DP ranks and document the supported configuration.

## Accuracy Tests

CPU routing and speculative-verification regressions: **16 passed**.

```bash
python -m pytest -q \
  test/registered/unit/lora/test_triton_dp_attention_unit.py \
  test/registered/unit/lora/test_lora_spec_verify_batch_info.py
```

These check routing metadata; end-to-end model accuracy was not rerun for this update.

## Speed Tests and Profiling

Not rerun for this update.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #34793736108](https://github.com/sgl-project/sglang/actions/runs/34793736108)
Latest PR Test (Extra): :white_check_mark: [Run #34793735848](https://github.com/sgl-project/sglang/actions/runs/34793735848)
Latest PR Test (AMD ROCm 10): :x: [Run #34793736079](https://github.com/sgl-project/sglang/actions/runs/34793736079)

-Robot

### CODE REVIEW

File: python/sglang/srt/lora/backend/base_backend.py [hunk 1]
```diff
@@ -52,6 +58,19 @@ def reset_batch_state(self):
...
+    def get_batch_info(
+        self, layout: LoRABatchLayout | None = None
+    ) -> Optional[LoRABatchInfo]:
+        """Return routing metadata for the requested token layout.
+
+        Backends that support multiple layouts override this method.
+        """
+        return self.batch_info
+
+    def prepare_global_lora_batch(self, forward_batch: ForwardBatch) -> None:
```
Comment (Pranshu258) [inline python/sglang/srt/lora/backend/base_backend.py:70]: If a future backend sets:

supports_dp_attention = True

but forgets to override the method, the server could proceed without constructing global routing metadata.

Comment (yushengsu-thu) [reply python/sglang/srt/lora/backend/base_backend.py:73]: fixed

**Review by yushengsu-thu — APPROVED** [review]

**Comment by yushengsu-thu** [comment]
| Item | Value |
|---|---|
| Model | `deepseek-ai/DeepSeek-V3.2-Exp` (FP8, 61 layers, 128 heads, MLA `kv_lora_rank` 512, DSA `index_topk` 2048, 256 routed experts) |
| LoRA adapter | rank 16, alpha 32, targets `q_a_proj`, `kv_a_proj_with_mqa`, `q_b_proj`, `o_proj`; pinned; Triton LoRA backend |
| Parallelism | no DPA: `--tp 8`; DPA: `--tp 8 --dp 8 --enable-dp-attention` (attn_tp = 1) |
| Memory | default `--mem-fraction-static` (auto-sized KV cache, so KV capacity reflects the memory budget) |
| Workload | `sglang.bench_serving`, random dataset, 512 requests, 4096 input / 512 output tokens, max concurrency 256, request rate ∞ |
| Software | `lmsysorg/sglang:dev` nightly (main `fa663e72`, torch 2.13.0+cu130, sglang-kernel 0.4.6.post1) with the PR branch (`fe555b9efd`) installed in editable mode |
| Validity | both runs started with all 8 GPUs empty (138 GB free at weight load) and had no foreign GPU process during the benchmark |

**Results (LoRA enabled in both columns)**

| Metric | LoRA, TP8 (no DPA) | LoRA, DP8 + DPA | Change |
|---|---|---|---|
| Request throughput (req/s) | 1.95 | **3.67** | **+88% (1.88×)** |
| Output token throughput (tok/s) | 997 | **1,877** | +88% |
| Total token throughput (tok/s) | 8,971 | 16,890 | +88% |
| Median TTFT (ms) | 76,387 | **23,743** | **−69%** |
| P99 TTFT (ms) | 121,792 | 46,149 | −62% |
| Median TPOT (ms) | 81.2 | 91.0 | +12% |
| Median E2E latency (ms) | 122,767 | 69,781 | −43% |
| Benchmark duration (s) | 263 | 140 | −47% |
| KV cache capacity (tokens) | 569,024 total, replicated on every GPU | 165,248 per rank = **1,321,984 total** | **2.32×** |
| KV cache memory (GB / GPU) | 41.5 | 12.1 | −29.4 |
| Model weights (GB / GPU) | 80.6 | 92.3 | +11.7 (attention weights replicated per rank) |
| GPU memory, peak (GB / GPU) | 139.9 | **119.0** | **−20.9** |
| Requests retracted during the run | 81 | 1 | |
| Chunked prefill size | 8192 | 1024 per rank (auto-divided by dp) | |

**Comment by hnyls2002** [comment]
Canceled the CI runs as H20 has too long a queuing time; this test can be skipped (I am trimming H20 tests)

@erikwijmans @yushengsu-thu

File: python/sglang/srt/layers/communicator.py [hunk 2]
```diff
@@ -654,6 +655,23 @@ def _post_attn_residual_is_read_only(self, residual: torch.Tensor) -> bool:
             and apply_flashinfer_allreduce_fusion(residual.shape[0])
         )
 
+    def publish_attn_lora_layout(self) -> None:
+        """Attention consumes the DP-local token batch."""
+        if get_parallel().enable_dp_attention:
```
Comment (Fridge003) [inline python/sglang/srt/layers/communicator.py:660]: Can we skip this function when lora is disabled

File: python/sglang/srt/layers/communicator.py [hunk 3]
```diff
@@ -654,6 +655,23 @@ def _post_attn_residual_is_read_only(self, residual: torch.Tensor) -> bool:
...
+            get_forward().set("lora_batch_layout", LoRABatchLayout.DP_LOCAL)
+
+    def publish_mlp_lora_layout(self) -> None:
+        """The MLP consumes the TP-global batch only after a FULL DP gather."""
```
Comment (Fridge003) [inline python/sglang/srt/layers/communicator.py:664]: Can we skip this function when lora is disabled

File: python/sglang/srt/layers/communicator.py [hunk 4]
```diff
@@ -654,6 +655,23 @@ def _post_attn_residual_is_read_only(self, residual: torch.Tensor) -> bool:
...
+        if get_parallel().enable_dp_attention:
+            get_forward().set(
+                "lora_batch_layout",
+                (
+                    LoRABatchLayout.TP_GLOBAL
+                    if self.layer_scatter_modes.mlp_mode is ScatterMode.FULL
+                    else LoRABatchLayout.DP_LOCAL
```
Comment (Fridge003) [inline python/sglang/srt/layers/communicator.py:671]: When 1 < dp_size < tp_size, the layout is neither LoRABatchLayout.DP_LOCAL nor LoRABatchLayout.TP_GLOBAL. Do we want to ban this case?

File: python/sglang/srt/layers/logits_processor.py [hunk 5]
```diff
@@ -828,7 +833,13 @@ def _get_logits(
             _trace_e2e_logits("pre_lm_head_sync_returned")
 
         _trace_e2e_logits("lm_head_enter", hidden_shape=tuple(hidden_states.shape))
-        logits = self._compute_lm_head(hidden_states, lm_head, embedding_bias)
+        lora_batch_layout = (
+            LoRABatchLayout.TP_GLOBAL
+            if self.do_tensor_parallel_all_gather_dp_attn
+            else LoRABatchLayout.DP_LOCAL
```
Comment (Fridge003) [inline python/sglang/srt/layers/logits_processor.py:839]: Same, what if 1 < dp_size < tp_size

File: python/sglang/srt/layers/communicator.py (hunk 2 above)
Comment (yushengsu-thu) [reply python/sglang/srt/layers/communicator.py]: Done in [05ab613e65](https://github.com/sgl-project/sglang/pull/36389/commits/05ab613e65e25ce2db94f80e4ced4f5b259a61d1): `LayerCommunicator` decides once at construction whether to publish the LoRA layout (DP attention with LoRA enabled), so both methods are no-ops without LoRA.

File: python/sglang/srt/layers/communicator.py (hunk 3 above)
Comment (yushengsu-thu) [reply python/sglang/srt/layers/communicator.py]: Done in [05ab613e65](https://github.com/sgl-project/sglang/pull/36389/commits/05ab613e65e25ce2db94f80e4ced4f5b259a61d1): `LayerCommunicator` decides once at construction whether to publish the LoRA layout (DP attention with LoRA enabled), so both methods are no-ops without LoRA.

File: python/sglang/srt/layers/communicator.py (hunk 4 above)
Comment (yushengsu-thu) [reply python/sglang/srt/layers/communicator.py]: Good catch. With 1 < dp_size < tp_size (or attention CP), SCATTERED/MOE_FULL MLP modes and attention-TP input scattering give a layout neither value describes. Banned in [17ec44758e](https://github.com/sgl-project/sglang/pull/36389/commits/17ec44758e13c2d0dafea891298b1f64fc7f262d): `LoRAManager` rejects DP attention unless `--dp-size` equals `--tp-size`; documented and unit-tested.

File: python/sglang/srt/layers/logits_processor.py (hunk 5 above)
Comment (yushengsu-thu) [reply python/sglang/srt/layers/logits_processor.py]: Covered by the same check in [17ec44758e](https://github.com/sgl-project/sglang/pull/36389/commits/17ec44758e13c2d0dafea891298b1f64fc7f262d). The LM head only sees DP-local tokens or the DP-gathered batch, and LoRA with DP attention now requires dp_size == tp_size.
